### ISSUE TITLE
"offset" is not allowed as identifier

### ISSUE THREAD

**Issue by wds15** [stan-dev/stanc3#453]
The variable name "offset" makes stanc3 reject valid Stan programs.

Example:

```stan
data {
  int<lower=0> N;
  int<lower=0,upper=1> y[N];
}
transformed data {
  int offset = 10;
}
parameters {
  real<lower=0,upper=1> theta;
}
model {
  theta ~ beta(1,1);  // uniform prior on interval 0,1
  y ~ bernoulli(theta);
}
```

gives me

```
--- Translating Stan model to C++ code ---
bin/stanc  --o=examples/bernoulli/bernoulli.hpp examples/bernoulli/bernoulli.stan

Syntax error in 'examples/bernoulli/bernoulli.stan', line 6, column 2 to column 5, parsing error:
   -------------------------------------------------
     4:  }
     5:  transformed data {
     6:    int offset = 10;
              ^
     7:  }
     8:  parameters {
   -------------------------------------------------

Expected range constraint or identifier as part of top-level variable declaration.



--- Translating Stan model to C++ code ---
bin/stanc  --o=examples/bernoulli/bernoulli.hpp examples/bernoulli/bernoulli.stan

Syntax error in 'examples/bernoulli/bernoulli.stan', line 6, column 2 to column 5, parsing error:
   -------------------------------------------------
     4:  }
     5:  transformed data {
     6:    int offset = 10;
              ^
     7:  }
     8:  parameters {
   -------------------------------------------------

Expected range constraint or identifier as part of top-level variable declaration.
```

This is a bit unfortunate as "offset" is used by brms, for example...

EDIT: this is the released cmdstan 2.21.1 on Mac (but same problem on Linux)

**Comment by avehtari** [comment]
Release notes should have included new reserved keywords in stanc3.
For offset see https://github.com/stan-dev/stanc3/issues/21

**Comment by avehtari** [comment]
See also discussion at https://discourse.mc-stan.org/t/list-of-stanc3-new-reserved-keywords/12948

**Comment by avehtari** [comment]
@mcol lists `target`, `lower`, `upper`, `offset` and `multiplier` as new reserved keywords, but is not certain whether this is complete list https://discourse.mc-stan.org/t/list-of-stanc3-new-reserved-keywords/12948/5

**Comment by wds15** [comment]
Ok... this is not a bug, but we clearly need better doc. Moreover, the error message is not useful for users. That should be really improved, unless I am overlooking something.

**Comment by rok-cesnovar** [comment]
The error it prints is the same as if you would define any of the following.

```stan
real break;
real for;
real vector;
```
These errors happen as part of the syntax check while most of the rest are part of the semantic check and happen later and give out the following error (note the "Semantic error" at the beginning):

```
 Semantic error in 'nullptr.stan', line 2, column 7 to column 14:
    -------------------------------------------------
      1:  data {
      2:    real nullptr;
                 ^
      3:  }
      4:  model {
    -------------------------------------------------
 
 Identifier 'nullptr' clashes with reserved keyword.
```

The error message for the syntax error is simple to change. Maybe we change it to: 

`A valid identifier expected after type in top-level variable declaration. The parser found no identifier or an invalid identifier (e.g. reserved keyword).`

I am not sure on what are the rules for an error message.

**Comment by avehtari** [comment]
Could the parser actually know that it is reserved keyword and the message could be
```
A valid identifier expected after type in top-level variable declaration. The parser found a reserved keyword which cannot be used as an identifier.
```

**Comment by rok-cesnovar** [comment]
That question is a bit above my "paygrade" for now. Tagging @VMatthijs who probably knows more.

In stanc3 tests this error is only found with keywords like I show above. I do also however get it with 

```stan
data {
  real ;
}
model {
}
```

**Comment by bob-carpenter** [comment]
> On Feb 4, 2020, at 12:40 PM, Aki Vehtari <notifications@github.com> wrote:
> 
> Could the parser actually know that it is reserved keyword and the message could be
> 
> A valid identifier expected after type in top-level variable declaration. The parser found a reserved keyword which cannot be used as an identifier.

How about

  Expected identifier, but found reserved keyword 'nullptr'

That would work in all the contexts where a reserved word is used as an identifier.  Otherwise, it's going to need rules for variable declarations, function arguments, uses in expressions, etc., which might be a lot trickier to write given the context required.

**Comment by rybern** [comment]
This was closed by #457
